cluster/addons/pod-resizer/src/nanny_lib.go

```diff
@@ -0,0 +1,81 @@
+package src
+
+import (
+	"time"
+
+	api "k8s.io/kubernetes/pkg/api/v1"
+
+	log "github.com/golang/glog"
+
+	inf "speter.net/go/exp/math/dec/inf"
+)
+
+// checkResource determines whether a specific resource needs to be over-written.
+func checkResource(threshold int64, actual, expected api.ResourceList, res api.ResourceName) bool {
+	val, ok := actual[res]
+	expVal, expOk := expected[res]
+	if ok != expOk {
+		return true
+	}
+	if !ok && !expOk {
+		return false
+	}
+	q := new(inf.Dec).QuoRound(val.Amount, expVal.Amount, 2, inf.RoundDown)
+	lower := inf.NewDec(100-threshold, 2)
+	upper := inf.NewDec(100+threshold, 2)
+	if q.Cmp(lower) == -1 || q.Cmp(upper) == 1 {
+		return true
+	}
+	return false
+}
+
+// shouldOverwriteResources determines if we should over-write the container's resource limits. We'll over-write the resource limits if the limited resources are different, or if any limit is violated by a threshold %.
+func shouldOverwriteResources(threshold int64, limits, reqs, expLimits, expReqs api.ResourceList) bool {
+	return checkResource(threshold, limits, expLimits, api.ResourceCPU) ||
+		checkResource(threshold, limits, expLimits, api.ResourceMemory) ||
+		checkResource(threshold, limits, expLimits, api.ResourceStorage) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceCPU) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceMemory) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceStorage)
+}
+
+func PollApiServer(k8s KubernetesClient, est ResourceEstimator, contName string, pollPeriod time.Duration, threshold uint64) {
+	for i := 0; true; i++ {
+		if i != 0 {
+			// Sleep for the poll period.
+			time.Sleep(pollPeriod)
+		}
+
+		// Query the apiserver for the number of nodes.
+		num, err := k8s.CountNodes()
+		if err != nil {
+			log.Error(err)
+			continue
```

Comment:
Would it be worth doing a log.Fatal if there are a large number of consecutive errors? The question being whether we'd be more likely to notice a restarting addon than an addon that is continually logging errors.